### ISSUE TITLE
limit action value of a continuous action

### ISSUE THREAD

**Issue by TheChrizZly** [Unity-Technologies/ml-agents#319]
Hello, 
I have an issue with my ml agents always using too high values in a continuous action.
I have an action which uses values from -1 to 1 as Inputs. This is the only action unity can do. 
But when I log the Input unity always tries values like 520,630,-580 etc.
How can I limit the Input value?
I looked in Unity itself and in the ppo implementation of ML-agents nowhere finding an obvious option to do it.

Thanks in advance

**Comment by MarcoMeter** [comment]
Hi @TheChrizZly 
simply clamp the action's value in AgentStep():
`Mathf.Clamp(action[0], -1, 1)`

**Comment by TheChrizZly** [comment]
Hi @MarcoMeter
Thanks for the answer.
But than I would have the problem that the PPO would still send high values. I would just crop them.
It would take like endless time to train as it would still try to send all the high values. Cannot I just make it more efficient and let the ppo just use values a from -1 to 1?

**Comment by MarcoMeter** [comment]
I just can say that many related projects utilize the clamping function right before the action is put to life. So far, i didn't find any explanation which gives further reasoning to the point that you make.

**Comment by vincentpierre** [comment]
One thing that people have used is a form of penalty for high values of actions. Think of it as an energy cost. If for the network sending an action value of 1 has the same result as sending 100, then the network would not know which is best. Maybe put a penalty for when the absolute value of the action is greater than 1 before clamping it. It does not need to have a high multiplier. I have not tested it but I would try: 
```csharp
reward -= 0.01f*(Mathf.Max(Mathf.Abs(action[0]), 1f) - 1f)
Mathf.Clamp(action[0], -1, 1)
```
If the action is between -1 and 1, the reward would not change. If it is out of range, the agent will receive a penalty proportional to the action value.

**Comment by MarcoMeter** [comment]
I started to play a little around within this context. A very simple example for the need of fine-grained outputs can be illustrated by choosing a direction to move to.

![env](https://user-images.githubusercontent.com/6996955/37817124-ead15f9c-2e75-11e8-86aa-5a7e71786df8.png)
The above environment features an agent (red), which is rewarded for collecting the target (green). Before, I utilized two continuous actions for moving horizontally or vertically. The cumbersome problem of this is the stair-like movement behavior. The desired result is to let the agent learn the perfect direction, which is the closest way to the target in the end.

Now, I'm trying to just use one continuous action for specifying an angle, which indicates the moving direction. Here is the code snippet, which shall accomplish carrying out such actions:

```
float rotationAngle = Mathf.Clamp(vectorAction[0], -1, 1) * 180;
Vector3 circumferencePoint = new Vector3((Mathf.Cos(rotationAngle * Mathf.Deg2Rad)),
                                                        (Mathf.Sin(rotationAngle * Mathf.Deg2Rad)),
                                                        0);
_rigidbody.velocity = circumferencePoint.normalized * _speed;
```

Basically, the angle can be used to derive a position from the unit circle, which serves as moving direction. During training, the agent moves left most of the time, because of action values exceeding the range +1 and -1. So at 0° (-1) or 360° (1) the direction is left (see below).

![test](https://user-images.githubusercontent.com/6996955/37817837-5d848d32-2e78-11e8-9049-b197b65d7e52.gif)

I'll let you know how things turn out after messing with rewards and other stuff concerning this example.
